### PR TITLE
Allow more intra-L0 compaction when L0 is small

### PR DESCRIPTION
**Pull request by cbi42**
Summary: introduce a new option `intra_l0_compaction_size` to allow more intra-L0 compaction when total L0 size is under a threshold. This option applies only to leveled compaction. It is enabled by default and set to `max_bytes_for_level_base / max_bytes_for_level_multiplier` only for atomic_flush users. When atomic_flush=true, it is more likely that some CF's total L0 size is small when it's eligible for compaction. This option aims to reduce write amplification in this case.

Test plan:
- new unit test
- benchmark: 
```
TEST_TMPDIR=/dev/shm ./db_bench --benchmarks=fillrandom --write_buffer_size=51200 --max_bytes_for_level_base=5242880 --level0_file_num_compaction_trigger=4 --statistics=1

main:
fillrandom   :     234.499 micros/op 4264 ops/sec 234.499 seconds 1000000 operations;    0.5 MB/s
rocksdb.compact.read.bytes COUNT : 1490756235
rocksdb.compact.write.bytes COUNT : 1469056734
rocksdb.flush.write.bytes COUNT : 71099011

branch:
fillrandom   :     128.494 micros/op 7782 ops/sec 128.494 seconds 1000000 operations;    0.9 MB/s
rocksdb.compact.read.bytes COUNT : 807474156
rocksdb.compact.write.bytes COUNT : 781977610
rocksdb.flush.write.bytes COUNT : 71098785
```

### CODE REVIEW

**Comment by hx235** [comment]
I happened to notice https://github.com/facebook/rocksdb/pull/10865 - are these related?

**Comment by cbi42** [comment]
> I happened to notice #10865 - are these related?

Yes, this is following https://github.com/facebook/rocksdb/pull/10865#issuecomment-1767177727 and will solve this same problem.

File: include/rocksdb/advanced_options.h [hunk 1]
```diff
@@ -1242,6 +1242,24 @@ struct AdvancedColumnFamilyOptions {
...
+  //
+  // UINT64_MAX - 1 (0xfffffffffffffffe) is a special flag to allow RocksDB
+  // to pick default.
+  // 0 means disabled.
+  //
+  // Default:
+  //  when atomic_flush = 0: 0
+  //  when atomic_flush=1:
+  //    max_bytes_for_level_base / max_bytes_for_level_multiplier
+  uint64_t intra_l0_compaction_size = 0xfffffffffffffffe;
```
Comment (ajkr) [inline include/rocksdb/advanced_options.h:1261]: I don't know if we need a new option. Maybe.

What if the threshold is determined based on the actual size(Lbase)? For example, size(L0) < size(Lbase) / max_bytes_for_level_multiplier. Maybe scale down the limit even more. The intuition is we don't want to run L0->Lbase compactions that are very inefficient for write-amp. Does that approach let unit tests pass?

Comment (cbi42) [reply include/rocksdb/advanced_options.h]: Good idea. Updated to remove the option and use `size(Lbase) / max(10, max_bytes_for_level_multiplier)` as the threshold. It turns out that no unit test change is needed.

**Comment by cbi42** [comment]
I also added some db_bench result to the PR summary.

**Review by ajkr — APPROVED** [review]
LGTM!

**Pull request closed without merging**